wisp/wisp.php

```diff
@@ -597,3 +680,155 @@ function wisp_ClientArea(array $params) {
+            $available_allocations[$ip][$port]['id'] = $allocation['attributes']['id'];
+        }
+    }
+    return $available_allocations;
+}
+
+function getPaginatedData($params,$url) {
+    /*
+        Makes a paginated API request and returns the response.
+    */
+
+    $results = array();
+    // Fetch and parse first page of data
+    $response = wisp_API($params, $url);
+    foreach($response['data'] as $key => $value) {
+        array_push($results, $value);
+    }
+    $current_page = $response['meta']['pagination']['current_page'];
+    $total_pages = $response['meta']['pagination']['total_pages'];
+    // Fetch and parse any remaining pages
+    while ($total_pages > $current_page) {
+        $next_page = intval($current_page)+1;
+        $response = wisp_API($params, $url.'?page='.$next_page);
+        foreach($response['data'] as $key => $value) {
+            array_push($results, $value);
+        }
+        $current_page = $response['meta']['pagination']['current_page'];
+        $total_pages = $response['meta']['pagination']['total_pages'];
+    }
+    return $results;
+}
+
+function findFreePorts(array $available_allocations, string $port_offsets) {
+    /*
+        This is the main logic that takes a list of available allocations
+        and the required offsets and then finds the first available set.
+        e.g. if port offsets +1 +2 and +4 are requested (format: 1,2,4)
+        we take each port one by one and check if <port number> + 1,
+        <port number> + 2 and <port number> + 4 are available.
+        If all requested port allocations are available, they are returned.
+
+        Inputs:
+        $available_allocations      This is the first port in the range to test.
+                                    All other ports will be checked based on the
+                                    required offset from the first.
+
+        $port_offsets               The json string of offsets from the first port that
+                                    are required for the server.
+
+        Outputs:
+        $ports_found                The array of ports that were found available, based on the offsets
+                                    the additional ports required.
+    */
+    
+    $port_offsets_array = json_decode($port_offsets, true);
+
+    // Iterate over available IP's
+    foreach($available_allocations as $ip_addr => $ports) {
+        $result = Array();
+        $result['status'] = false;
+        $main_allocation_id = "";
+        $main_allocation_port = "";
+        $additional_allocation_ids = Array();
+        $additional_allocation_ports = Array();
+
+        // Iterate over Ports
+        logModuleCall("WISP-WHMCS", "Checking IP: ".$ip_addr, "", "");
+        foreach($ports as $port => $portDetails) {
+            $main_allocation_id = $portDetails['id'];
+            $main_allocation_port = $port;
+            $found_all = true;
+            foreach($port_offsets_array as $key => $port_offset) {
```

Comment:
Foreach key and value are swapped
There the key (which you use as port offset) is used as parameter name and the value (which you use as parameter name) is used as port offset.